### ISSUE TITLE
support query ref model via populate-virtuals

### ISSUE THREAD

**Issue by mulder1628** [NathanAdhitya/express-mongoose-ra-json-server#5]
Hi Nathan,

I have a similar set up where my posts model has a ref to users model

```
┌──────────────┐       ┌────────────────┐
│ posts        │       │ users          │
│--------------│       │----------------│
│ id           │   ┌───│ id             │
│ user_id      │╾──┘   │ name           │
│ title        │       │ date_of_birth  │
│ published_at │       └────────────────┘
└──────────────┘

```

on the frontend (ra), i set up a filter in my post collection

```
const postFilters = [
  <TextInput source="q" label="Search" alwaysOn />,
]

```
and I want to be able this q to filter all the posts to the (semi)matching user name (not the objectId of user).

Here is my set up in the backend:
```
const raPostOptions = {
  q: ["user_id"],
};

app.use("/posts", protect, raExpressMongoose(PostModel, raPostOptions));

```
I read about the populate-virtuals in mongoose, is there a way to configure express-mongoose-ra-json-server to query against name in the ref users model from the posts model.

Here is the sample for such query from stackoverflow:
https://stackoverflow.com/questions/59138481/mongoose-find-by-reference-field

Thanks for your reply in advance!

**Comment by NathanAdhitya** [comment]
Hello, this is doable entirely on the frontend.
Consider checking the demo project to see how it's done. 

Relevant files are here:
https://github.com/NathanAdhitya/express-mongoose-ra-json-server-demo/blob/3fb20b11b979b754524bf145e7d88ea4cfbd8569/packages/frontend/src/app/resources/post.resource.tsx

https://github.com/NathanAdhitya/express-mongoose-ra-json-server-demo/blob/3fb20b11b979b754524bf145e7d88ea4cfbd8569/packages/backend/src/api/index.ts

If you do have better ideas, then let me know.

**Comment by mulder1628** [comment]
Hi Nathan,

Thanks for the prompt reply.  You have misunderstood me.  I want to be able to do this using the q filter, not the ReferenceInput, as I have already spent the ReferenceInput quota for another field.  RA has a limit of 1 ReferenceInput (which I have been trying to get around w/o success), that's why I am exploring q filter (TextInput) to provide richer filter in the post collection.

Here is my complete postFilters (remove initially to avoid confusion.

```
const postFilters = [
  <TextInput source="q" label="Search" alwaysOn />,
  <ReferenceInput
    source="approver"
    label="Approved by"
    reference="approver"
    perPage={100}
    sort={{ field: "name", order: "ASC" }}
  >
]
```

Thanks again!

**Comment by NathanAdhitya** [comment]
Hello. I don't think RA has a limit of 1 reference input that you can use on a filter.
Check the new branch I have published that uses two Reference Inputs to filter out based on post writer and post reviewer.
https://github.com/NathanAdhitya/express-mongoose-ra-json-server-demo/tree/more-filters

Here are the filters:
```ts
const filters = [
  <TextInput key="q" source="q" label="Search" alwaysOn />,
  <ReferenceInput source="userId" reference="user" label="Writer">
    <AutocompleteInput optionText="name" optionValue="id" label="Writer" />
  </ReferenceInput>,
  <ReferenceInput source="reviewerUserId" reference="user" label="Reviewer">
    <AutocompleteInput optionText="name" optionValue="id" label="Reviewer" />
  </ReferenceInput>,
];
```

![image](https://user-images.githubusercontent.com/13374903/192132740-18150830-37e5-4ab3-a7b8-94e4ec9cebb5.png)
![image](https://user-images.githubusercontent.com/13374903/192132741-e18ce3df-d8d7-4e9a-8f50-e09ddec03b24.png)

**Comment by NathanAdhitya** [comment]
Additional information: With the current implementation, q only looks for fields inside the same document in the collection it was declared with.

So, to answer your main question "is there a way to configure express-mongoose-ra-json-server to query against name in the ref users model from the posts model" via q? Not with the current implementation.

If you wish to implement custom logic, you can supply your own custom behavior per route as asked in issue #2.

**Comment by mulder1628** [comment]
Hi Nathan,
Thanks for the reference to using multiple ReferenceInput in the filter.  I was confused when I was getting this error using the default AutocompleteInput component.

`<ReferenceInput> only accepts a single child`

I tried your code, and got my posts to support multiple ReferenceInputs, that makes having the q generic filter to support populate-virtuals unnecessary.

Btw, lots of functionalities to fiddle with in raExpressMongooseOptions, would be great if there are examples to get us started on using them.
 
Thanks again for your help!  
Closing this issue!
